kubernetes-client/src/main/java/io/fabric8/kubernetes/client/CustomResourceMetadata.java

```diff
@@ -0,0 +1,166 @@
+   * input to {@link Pluralize#toPlural(String)}.
+   *
+   * @param clazz the CustomResourceMetadata whose plural form we want to retrieve
+   * @return the plural form defined by the {@link Plural} annotation or a computed default value
+   */
+  static String getPlural(Class<? extends CustomResourceMetadata> clazz) {
+    final Plural fromAnnotation = clazz.getAnnotation(Plural.class);
+    return (fromAnnotation != null ? fromAnnotation.value().toLowerCase(Locale.ROOT)
+      : Pluralize.toPlural(getSingular(clazz)));
+  }
+
+  /**
+   * Retrieves the singular form associated with the specified CustomResourceMetadata as defined by the
+   * {@link Singular} annotation or computes a default value (lower-cased version of the value
+   * returned by {@link HasMetadata#getKind(Class)}) if the annotation is not present.
+   *
+   * @param clazz the CustomResourceMetadata whose singular form we want to retrieve
+   * @return the singular form defined by the {@link Singular} annotation or a computed default
+   * value
+   */
+  static String getSingular(Class<? extends CustomResourceMetadata> clazz) {
+    final Singular fromAnnotation = clazz.getAnnotation(Singular.class);
+    return (fromAnnotation != null ? fromAnnotation.value() : HasMetadata.getKind(clazz))
+      .toLowerCase(Locale.ROOT);
+  }
+
+  /**
+   * Computes the name of the Custom Resource Definition (CRD) associated with the specified
+   * CustomResourceMetadata. See https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/
+   * for more details.
+   *
+   * @param clazz the CustomResourceMetadata whose CRD name we want to compute
+   * @return the CRD name associated with the CustomResourceMetadata
+   */
+  static String getCRDName(Class<? extends CustomResourceMetadata> clazz) {
+    return getCRDName(getPlural(clazz), HasMetadata.getGroup(clazz));
+  }
+
+  static String getCRDName(String plural, String group) {
+    return plural + "." + group;
+  }
+
+  @JsonIgnore
+  default String getPlural() {
+    return getPlural(getClass());
+  }
+
+  @JsonIgnore
+  default String getSingular() {
+    return getSingular(getClass());
+  }
+
+
+  @JsonIgnore
+  default String getCRDName() {
+    return getCRDName(getClass());
+  }
+
+  /**
+   * Retrieves the scope that this CustomResource targets
+   *
+   * @return the scope that this CustomResource targets. Possible values are {@link #CLUSTER_SCOPE}
+   * or {@link #NAMESPACE_SCOPE}.
+   */
+  @JsonIgnore
+  default String getScope() {
+    return getScope(getClass());
+  }
+
+  @JsonIgnore
+  default String getGroup() {
+    return HasMetadata.getGroup(getClass());
+  }
+
+  @JsonIgnore
+  default String getVersion() {
+    return HasMetadata.getVersion(getClass());
+  }
+
+  @JsonIgnore
+  default boolean isServed() {
+    return getServed(getClass());
+  }
+
+  @JsonIgnore
+  default boolean isStorage() {
+    return getStorage(getClass());
+  }
+
+  S getSpec();
+
+  T getStatus();
```

Comment:
What is the use of having getSpec() and getStatus() here? it's not enough to have them in the abstract class?

I think having them here defeats the purpose of a "Metadata" type.